### ISSUE TITLE
Handling Refresh Token Expiry

### ISSUE THREAD

**Issue by james-ff** [aws-amplify/aws-sdk-ios#2959]
**Describe the bug**
We're seeing the following issue:

A user signs in successfully into our app via Cognito Custom Auth. After the refresh token expiry elapses (default of 30 days in our production environment, and 1 day In our staging environment), the user opens the app from background, is logged out (due to `.signedOutUserPoolsTokenInvalid`).
 
When they attempt to re-sign in with their phone or email as username (which we pass on to the signIn method as we do on all sign in attempts), we are seeing the following error: `invalidParameter(message: "Missing required parameter USERNAME")`. They cannot sign in until the app is closed from the app switcher and launched again.
 
We believe this issue is seen only within a concurrent session (app in foreground / background, but not closed), and the state within AWSMobileClient might not be cleaned fully on token expiry, leading to the subsequent sign in failing. (A possibly similar report here: https://github.com/aws-amplify/aws-sdk-ios/issues/1669) . 

**To Reproduce**
1. Sign in with Custom Auth (signIn followed by challengeResponse in confirmSignIn)
2. Leave App open but in the background for the time it takes the refresh token to expire (minimum one day)
3. Open App, attempt to sign in again, see error listed above

**Observed Behavior**
Subsequent sign in after token expiry with `.signedOutUserPoolsTokenInvalid` fails, until app is closed and re-opened from the home screen

**Expected Behavior**
aws session and / or other variables clear on `.signedOutUserPoolsTokenInvalid` state so that the next login can work in the same app session.

**Areas of the SDK you are using (AWSMobileClient, Cognito, Pinpoint, IoT, etc)?**
We use Custom Auth configuration with AWSMobileClient

**Environment(please complete the following information):**
 - SDK Version: 2.15.1
 - Dependency Manager: Cocoapods
 - Swift Version: 5
 - Xcode Version: 11.6

**Comment by james-ff** [comment]
For local testing of this issue the token expiry wait time can be bypassed with the following:
```
func invalidateSession(username: String) {
        let bundleID = Bundle.main.bundleIdentifier
        let keychain = AWSUICKeyChainStore(service: "\(bundleID!).\(AWSCognitoIdentityUserPool.self)")
        let namespace = "\(AWSMobileClient.default().userPoolClient!.userPoolConfiguration.clientId).\(username)"
        let key = "\(namespace).tokenExpiration"
        keychain.removeItem(forKey: key)
}
```
(this is from the testSignOutOnSessionExpiry function in the test files)

**Comment by palpatim** [comment]
@james-ff Thanks for the investigation and the details. We'll take a look at this and update when we have more information.

**Comment by james-ff** [comment]
We've managed to see, for manual tokenExpiry invalidation, some resolution. But this hasn't translated over to the real use case of refresh token expiry. It seems that the SDK handles this case (where tokenExpiry is not nil, and not valid) differently.

One of the things we did was to add a call to `getTokens` (if signed in) in `applicationWillEnterForeground`. Hoping to trigger the SDK to realise that the token is invalid as soon as possible. And moved notifications for refresh to trigger afterwards with `UIApplication.didBecomeActiveNotification`.

**Comment by royjit** [comment]
Can you please retest this. I tried replicating this issue and could not reproduce. I am suspecting that the issue is resolved with this fix - https://github.com/aws-amplify/aws-sdk-ios/pull/3198 .

Here is the steps I followed to reproduce the issue.

1. Created a user pool that allow to sign in using email and username
2. Kept the token expiry time to minimum
3. Created a user
4. SignedIn using custom auth `AWSMobileClient.default().signin` using email as the username.
5.  Checked whether I am getting tokens : `AWSMobileClient.default().getToken`
6. Waited for the session to expiry 
7. Signed in again `AWSMobileClient.default().signin` using email as the username.

**Comment by wooj2** [comment]
We are optimistically closing this issue as we believe we have released a fix.  Feel free to re-open if you have any questions, comments or concerns regarding this issue.

**Comment by jamesingham** [comment]
Are you sure this is resolved? I am using the latest SDK (v2.26.7) and if I do the following:

1. Sign in with AWSMobileClient.default().signIn() - signin success
2. Call AWSMobileClient.default().getTokens() - I can see all the tokens and expiry time in the callback
3. Wait until the refresh token expires (I currently have it set to 60 mins for testing)
4. Call AWSMobileClient.default().getTokens() again

Once the refresh token is expired, the completionHandler callback for getTokens() is never called. Instead I only see this in the logs:
`[Client] Updating selectors after delegate removal failed with: Error Domain=NSCocoaErrorDomain Code=4099 "The connection to service on pid 93 named com.apple.commcenter.coretelephony.xpc was invalidated from this process." UserInfo={NSDebugDescription=The connection to service on pid 93 named com.apple.commcenter.coretelephony.xpc was invalidated from this process.}
[connection] nw_endpoint_handler_set_adaptive_read_handler [C1.1.1 3.11.31.143:443 ready channel-flow (satisfied (Path is satisfied), viable, interface: en0, ipv4, dns)] unregister notification for read_timeout failed
[connection] nw_endpoint_handler_set_adaptive_write_handler [C1.1.1 3.11.31.143:443 ready channel-flow (satisfied (Path is satisfied), viable, interface: en0, ipv4, dns)] unregister notification for write_timeout failed`

**Comment by psoto-dev** [comment]
> Are you sure this is resolved? I am using the latest SDK (v2.26.7) and if I do the following:
> 
> 1. Sign in with AWSMobileClient.default().signIn() - signin success
> 2. Call AWSMobileClient.default().getTokens() - I can see all the tokens and expiry time in the callback
> 3. Wait until the refresh token expires (I currently have it set to 60 mins for testing)
> 4. Call AWSMobileClient.default().getTokens() again
> 
> Once the refresh token is expired, the completionHandler callback for getTokens() is never called. It seems to continue as soon as .signIn() has completed again however. (Tested on real hardware)


Sounds like intended behavior. Calling `getTokens()` after refresh token expires triggers a user state change notification of `signedOutUserPoolsTokenInvalid` or `signedOutFederatedTokensInvalid`. The completion handler will not be called until this state is handled (using either `signIn` or `releaseSignInWait` method).

From [UserState enum doc](https://aws-amplify.github.io/aws-sdk-ios/docs/reference/AWSMobileClient/Enums/UserState.html):
- `signedOutUserPoolsTokenInvalid`: The Cognito User Pools tokens have expired and the end-user needs to sign in again. You can complete the sign in process using the signIn method or by showing the drop in UI. If the end user is unable to sign in, call releaseSignInWait method to return an error to the calling API.

Refer to [Amplify Token Fetch and Refresh](https://docs.amplify.aws/sdk/auth/how-it-works/q/platform/ios/#token-fetch-and-refresh).
